src/ng/animate.js

```diff
@@ -416,8 +416,8 @@ var $AnimateProvider = ['$provide', /** @this */ function($provide) {
        * @param {Promise} animationPromise The animation promise that is returned when an animation is started.
        */
       cancel: function(runner) {
-        if (runner.end) {
-          runner.end();
+        if (runner.cancel) {
+          runner.cancel();
```

Comment:
So, the change with calling runner.cancel on cancel is not correct (at least not in this context), it just ensures the tests pass.
The problem is that in the ngSwitch tests, the animations are never completed with $$rAF flush. 
Howevr, previously unfinished leave animations are cancelled when we switch to a new element. This is a unique behavior of ngSwitch. Now cancel calls end, which executes everything inside the _doneCallbacks *without* waiting for an rAF (this might be an issue of its own). This wasn't a problem when using `then`, because the promise is resolved on the next tick. However, when using `done`, the _doneCallbacks are called synchronously, which means when there are two unfinished leave animations, the previousLeaveAnimations array is length 1 after the first animation is removed, but the second animation tries to remove the second element (now element 1).

with then and without flush
- switch 1: nothing
- switch 2: put both switch 1 leave animations into previousAnimation
  (they are not completed / removed, because not AF is flushed)
- switch true:
-- both 1+2 are identified as previousLeaveAnimations and are cancelled, which
calls the doneCallbacks (without extra rAF) 
-- however, since we are using promises they are only invoked after the next tick
-- switch 2 leave animation is put into previousAnimation
  (it are not completed / removed, because no AF is flushed / digest is called)

with done and without flush
- switch 1: nothing
- switch 2: put both switch 1 leave animations into previousAnimation
  (they are not completed / removed, because not AF is flushed)
- switch true:
-- both 1+2 are identified as previousLeaveAnimations and are cancelled, which
calls the doneCallbacks (without extra rAF) 
-- since we are not using promises they are immediately invoked, which means the first stored
previousLeaveAnimation is removed from the array, which means the second stored animation tries
to access index 1 which doesn' exist anymore

with done && animate.cancel =>
- done callbacks are called with false, which means the stored previousAnimations are never removed